plugins/qlcplus_4.12.1/import.js

```diff
@@ -0,0 +1,913 @@
+
+  fixture.meta = {
+    authors: authors,
+    createDate: timestamp,
+    lastModifyDate: timestamp,
+    importPlugin: {
+      plugin: `qlcplus_4.12.1`,
+      date: timestamp,
+      comment: `created by ${qlcPlusFixture.Creator[0].Name[0]} (version ${qlcPlusFixture.Creator[0].Version[0]})`
+    }
+  };
+
+  if (!(`Mode` in qlcPlusFixture)) {
+    qlcPlusFixture.Mode = [];
+  }
+  if (!(`Channel` in qlcPlusFixture)) {
+    qlcPlusFixture.Channel = [];
+  }
+
+  addOflFixturePhysical(fixture, qlcPlusFixture);
+
+  fixture.matrix = getOflMatrix(qlcPlusFixture);
+  fixture.wheels = getOflWheels(qlcPlusFixture);
+  fixture.availableChannels = {};
+  fixture.templateChannels = {};
+
+  qlcPlusFixture.Channel.forEach(channel => addOflChannel(fixture, channel, qlcPlusFixture));
+
+  mergeFineChannels(fixture, qlcPlusFixture, out.warnings[fixKey]);
+
+  fixture.modes = qlcPlusFixture.Mode.map(mode => getOflMode(mode, fixture.physical, out.warnings[fixKey]));
+
+  addSwitchingChannels(fixture, qlcPlusFixture);
+
+  cleanUpFixture(fixture, qlcPlusFixture);
+
+  out.fixtures[fixKey] = fixture;
+
+  return out;
+};
+
+/**
+ * @param {object} qlcPlusFixture The QLC+ fixture object.
+ * @returns {array.<string>} The OFL fixture categories.
+ */
+function getOflCategories(qlcPlusFixture) {
+  const category = qlcPlusFixture.Type[0];
+
+  if (category.startsWith(`LED Bar`)) {
+    return [`Pixel Bar`];
+  }
+
+  return [category];
+}
+
+/**
+ * Adds a global physical object to the OFL fixture object, if necessary.
+ * @param {object} fixture The OFL fixture object.
+ * @param {object} qlcPlusFixture The QLC+ fixture object.
+ */
+function addOflFixturePhysical(fixture, qlcPlusFixture) {
+  const allModesHavePhysical = qlcPlusFixture.Mode.every(mode => `Physical` in mode);
+  const firstPhysicalMode = qlcPlusFixture.Mode.find(mode => `Physical` in mode);
+  const hasGlobalPhysical = `Physical` in qlcPlusFixture;
+
+  if (!hasGlobalPhysical && (allModesHavePhysical || !firstPhysicalMode)) {
+    return;
+  }
+
+  fixture.physical = getOflPhysical(hasGlobalPhysical ? qlcPlusFixture.Physical[0] : firstPhysicalMode.Physical[0]);
+
+  if (qlcPlusFixture.Type[0] === `LED Bar (Pixels)`) {
+    fixture.physical.matrixPixels = {
+      spacing: [0, 0, 0]
+    };
+  }
+}
+
+/**
+ * @param {object} qlcPlusFixture The QLC+ fixture object.
+ * @returns {object} The OFL matrix object (may be empty).
+ */
+function getOflMatrix(qlcPlusFixture) {
+  const matrix = {};
+
+  const physicalLayouts = qlcPlusFixture.Mode.concat(qlcPlusFixture)
+    .filter(obj => `Physical` in obj && `Layout` in obj.Physical[0])
+    .map(obj => obj.Physical[0].Layout[0]);
+
+  if (physicalLayouts) {
+    const maxWidth = Math.max(...physicalLayouts.map(layout => parseInt(layout.$.Width)));
+    const maxHeight = Math.max(...physicalLayouts.map(layout => parseInt(layout.$.Height)));
+
+    matrix.pixelCount = [maxWidth, maxHeight, 1];
+  }
+
+  return matrix;
+}
+
+const slotTypeFunctions = {
```

Comment:
Maybe move this to directly above the `getSlots` method where this object is actually used.

Reply:
I'd like to keep it in the global scope because it's not dependent on any local state. Also, it does not have to be recreated whenever the `getSlots` function is called. It's already directly above the `getOflWheels` function.